### PR TITLE
Fix InexactError() on windows when trying to mmap large (>2GB) files.…

### LINKED ISSUE #13560: Mmap.mmap throws InexactError for filesize > 2Gb on Windows
Following code on Windows 7 64bit, Julia 0.4.0 shows the error:
```
julia> versioninfo()
Julia Version 0.4.0
Commit 0ff703b* (2015-10-08 06:20 UTC)
Platform Info:
  System: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-3520M CPU @ 2.90GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Sandybridge)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
julia> Mmap.mmap(open(tempname(),"w+"), Vector{UInt8}, 1024*1024*3048,0)
ERROR: InexactError()
 in mmap at mmap.jl:126
```
I think this origins in the usage of a wrong type `Cint` for a `ccall` in mmap.il:126
```
handle = create ? ccall(:CreateFileMappingW, stdcall, Ptr{Void}, (Cptrdiff_t, Ptr{Void}, Cint, Cint, Cint, Cwstring),
						file_desc, C_NULL, readonly ? PAGE_READONLY : PAGE_READWRITE, szfile >> 32, szfile & typemax(UInt32), name) :
				  ccall(:OpenFileMappingW, stdcall, Ptr{Void}, (Cint, Cint, Cwstring),
					readonly ? FILE_MAP_READ : FILE_MAP_WRITE, true, name)
```
Changing param types from `Cint` to `Cuint` (DWORD as specified for calls `CreateFileMappingW` and `OpenFileMappingW` in https://msdn.microsoft.com/en-us/library/windows/desktop/aa366763%28v=vs.85%29.aspx) lets mmap handle files with a size > 2Gb:
```
handle = create ? ccall(:CreateFileMappingW, stdcall, Ptr{Void}, (Cptrdiff_t, Ptr{Void}, Cuint, Cuint, Cuint, Cwstring),
						file_desc, C_NULL, readonly ? PAGE_READONLY : PAGE_READWRITE, szfile >> 32, szfile & typemax(UInt32), name) :
				  ccall(:OpenFileMappingW, stdcall, Ptr{Void}, (Cuint, Cuint, Cwstring),
					readonly ? FILE_MAP_READ : FILE_MAP_WRITE, true, name)
```

Or did I miss something?

### PR DESCRIPTION
**Pull request by quinnj**
… Fixes #13560

### CODE REVIEW

File: base/mmap.jl [hunk 1]
```diff
@@ -123,10 +123,10 @@ function mmap{T,N}(io::IO,
         name, readonly, create = settings(io)
         szfile = convert(Csize_t, len + offset)
         readonly && szfile > filesize(io) && throw(ArgumentError("unable to increase file size to $szfile due to read-only permissions"))
-        handle = create ? ccall(:CreateFileMappingW, stdcall, Ptr{Void}, (Cptrdiff_t, Ptr{Void}, Cint, Cint, Cint, Cwstring),
-                                file_desc, C_NULL, readonly ? PAGE_READONLY : PAGE_READWRITE, szfile >> 32, szfile & typemax(UInt32), name) :
-                          ccall(:OpenFileMappingW, stdcall, Ptr{Void}, (Cint, Cint, Cwstring),
-                            readonly ? FILE_MAP_READ : FILE_MAP_WRITE, true, name)
+        handle = create ? ccall(:CreateFileMappingW, stdcall, Ptr{Void}, (Cptrdiff_t, Ptr{Void}, Cuint, Cuint, Cuint, Cwstring),
+                        file_desc, C_NULL, readonly ? PAGE_READONLY : PAGE_READWRITE, szfile >> 32, szfile & typemax(UInt32), name) :
+                  ccall(:OpenFileMappingW, stdcall, Ptr{Void}, (Cuint, Cuint, Cwstring),
```
Comment (tkelman) [inline base/mmap.jl]: indentation is a little odd here, make sure it's all spaces instead of tabs?

Comment (quinnj) [inline base/mmap.jl]: I ran `make check-whitespace` and it seemed to pass. That would catch it, right?

Comment (tkelman) [inline base/mmap.jl]: check-whitespace doesn't enforce tabs or look at the initial indent of anything, just trailing whitespace at the end of a line

**Comment by quinnj** [comment]
Whitespace should be more consistent now.

File: base/mmap.jl [hunk 2]
```diff
@@ -123,10 +123,10 @@ function mmap{T,N}(io::IO,
         name, readonly, create = settings(io)
         szfile = convert(Csize_t, len + offset)
         readonly && szfile > filesize(io) && throw(ArgumentError("unable to increase file size to $szfile due to read-only permissions"))
-        handle = create ? ccall(:CreateFileMappingW, stdcall, Ptr{Void}, (Cptrdiff_t, Ptr{Void}, Cint, Cint, Cint, Cwstring),
+        handle = create ? ccall(:CreateFileMappingW, stdcall, Ptr{Void}, (Cptrdiff_t, Ptr{Void}, Cuint, Cuint, Cuint, Cwstring),
                                 file_desc, C_NULL, readonly ? PAGE_READONLY : PAGE_READWRITE, szfile >> 32, szfile & typemax(UInt32), name) :
-                          ccall(:OpenFileMappingW, stdcall, Ptr{Void}, (Cint, Cint, Cwstring),
-                            readonly ? FILE_MAP_READ : FILE_MAP_WRITE, true, name)
+                          ccall(:OpenFileMappingW, stdcall, Ptr{Void}, (Cuint, Cuint, Cwstring),
```
Comment (tkelman) [inline base/mmap.jl]: I think the second input here should stay `Cint`, ref https://msdn.microsoft.com/en-us/library/windows/desktop/aa366791(v=vs.85).aspx and https://msdn.microsoft.com/en-us/library/windows/desktop/aa383751(v=vs.85).aspx

**Comment by quinnj** [comment]
Ok, everything should be properly typed now.

File: base/mmap.jl [hunk 3]
```diff
@@ -121,14 +123,14 @@ function mmap{T,N}(io::IO,
...
-        handle = create ? ccall(:CreateFileMappingW, stdcall, Ptr{Void}, (Cptrdiff_t, Ptr{Void}, Cint, Cint, Cint, Cwstring),
+        handle = create ? ccall(:CreateFileMappingW, stdcall, Ptr{Void}, (Cptrdiff_t, Ptr{Void}, DWORD, DWORD, DWORD, Cwstring),
                                 file_desc, C_NULL, readonly ? PAGE_READONLY : PAGE_READWRITE, szfile >> 32, szfile & typemax(UInt32), name) :
-                          ccall(:OpenFileMappingW, stdcall, Ptr{Void}, (Cint, Cint, Cwstring),
-                            readonly ? FILE_MAP_READ : FILE_MAP_WRITE, true, name)
+                          ccall(:OpenFileMappingW, stdcall, Ptr{Void}, (DWORD, Cint, Cwstring),
+                                readonly ? FILE_MAP_READ : FILE_MAP_WRITE, true, name)
         handle == C_NULL && error("could not create file mapping: $(Libc.FormatMessage())")
-        ptr = ccall(:MapViewOfFile, stdcall, Ptr{Void}, (Ptr{Void}, Cint, Cint, Cint, Csize_t),
+        ptr = ccall(:MapViewOfFile, stdcall, Ptr{Void}, (Ptr{Void}, DWORD, DWORD, DWORD, DWORD),
```
Comment (tkelman) [inline base/mmap.jl]: 3 `DWORD`s and a `SIZE_T`? https://msdn.microsoft.com/en-us/library/windows/desktop/aa366761(v=vs.85).aspx

Comment (quinnj) [inline base/mmap.jl]: I was going off of: https://msdn.microsoft.com/en-us/library/aa914405.aspx.

Comment (tkelman) [inline base/mmap.jl]: careful
> Windows Mobile 6.5

apparently the mobile version of the winapi has different types. lovely.

Comment (quinnj) [inline base/mmap.jl]: Ah, good catch. I'll try to update tonight.

**Comment by tkelman** [comment]
I wonder if we could pick a package in which to check for the `APPVEYOR` environment variable, and when create a really big file to test this when it's set. AppVeyor lets you set scheduled builds so it could run once a day which is pretty useful for tracking.

**Comment by quinnj** [comment]
Not a bad idea. So how would that work? We'd create a `ExtraBaseTests.jl` package that would just house extra tests that would get run once a day on nightlies?

**Comment by tkelman** [comment]
Something like that, yeah.
@IainNZ has been accumulating a few perf-related things at https://github.com/IainNZ/BasePerfTests.jl which could also use the same treatment as long as they don't take too long to run.

**Pull request merged**